src/composition.rs

```diff
@@ -0,0 +1,649 @@
+                                                        _ => (),
+                                                    }
+                                                }
+                                            }
+                                        }
+                                    }
+                                })
+                            }
+                        });
+                }
+            }
+        }
+        true
+    }
+}
+
+/// Find the composition statement.
+///
+/// The composition statement is identified as follows: a `from` function that contains
+/// a yield with the name "_editor_composition".
+#[derive(Default)]
+struct CompositionStatementFinderVisitor {
+    statement: Option<ast::ExprStmt>,
+}
+
+impl<'a> ast::walk::Visitor<'a>
+    for CompositionStatementFinderVisitor
+{
+    fn visit(&mut self, node: ast::walk::Node<'a>) -> bool {
+        if self.statement.is_some() {
+            // If the statement was found, don't keep looking.
+            return false;
+        }
+
+        if let ast::walk::Node::ExprStmt(expr_statement) = node {
+            if let ast::Expression::PipeExpr(expr) =
+                &expr_statement.expression
+            {
+                if let ast::Expression::Identifier(identifier) =
+                    &expr.call.callee
+                {
+                    if identifier.name == "yield" {
+                        expr.call.arguments.iter().any(|argument| {
+                            if let ast::Expression::Object(object) = argument {
+                                object.properties.iter().any(|property| {
+                                    if let ast::PropertyKey::Identifier(key) = &property.key {
+                                        if key.name == "name" {
+                                            // Because we would have generated this statement, we'll always be able to
+                                            // assert the simplicity of the yield name, i.e. it won't be a deeply nested
+                                            // expression.
+                                            if let Some(ast::Expression::StringLit(literal)) = &property.value {
+                                                if literal.value == YIELD_IDENTIFIER {
+                                                    self.statement = Some(expr_statement.clone());
+                                                    return true;
+                                                }
+                                            }
+                                        }
+                                    }
+                                    false
+                                })
+                            } else {
+                                false
+                            }
+                        });
+                        return false;
+                    }
+                }
+            }
+        }
+        true
```

Comment:
Should it be `false` here since there is not a match for `ExprStmt`?

Reply:
_Maybe?_ I tend to just let it walk the entire AST unless I have a reason not to.